### PR TITLE
fix: core dump in macos running sql_cluster_test

### PR DESCRIPTION
**Pull request by aceforeverd**
1. fix a `sql_cluster_test` core in darwin: when there is no partition data for a table locally, the `DistributeWindowIterator::SeekToFirst` set `it_` still. This cores the program when user want to `DistributeWindowIterator::GetKey()` because `it_` is not null, so it try to get key data from `it_`
2. enable more last join and window union test for cluster/request mode
3. update codecov.yml
4. improve `toydb_run_engine` bootstrapping

### CODE REVIEW

**Review by zhanghaohit — APPROVED** [review]
LGTM

**Review by vagetablechicken — APPROVED** [review]

File: src/catalog/distribute_iterator.cc [hunk 1]
```diff
@@ -227,22 +227,30 @@ void DistributeWindowIterator::SeekToFirst() {
...
+            delete it;
         }
     }
     for (const auto& kv : tablet_clients_) {
         uint32_t count = 0;
-        cur_pid_ = kv.first;
-        kv_it_ = kv.second->Traverse(tid_, cur_pid_, index_name_, "", 0, FLAGS_traverse_cnt_limit, count);
-        if (kv_it_ && kv_it_->Valid()) {
-            response_vec_.emplace_back(kv_it_->GetResponse());
+        auto it = kv.second->Traverse(tid_, kv.first, index_name_, "", 0, FLAGS_traverse_cnt_limit, count);
```
Comment (dl239) [inline src/catalog/distribute_iterator.cc:228]: the `Seek` has the same problem. it's better to check it_ validation in `GetKey` and `GetValue` function

Comment (aceforeverd) [reply src/catalog/distribute_iterator.cc:228]: we should ensure only zero or one of `it_` and `kv_it_` is not null.
So the check here is whether `it_` and `kv_it_` both not null.
It is `it_` and `kv_it_` 's duty to check valid itself before `GetKey` or `GetValue`

Comment (aceforeverd) [reply src/catalog/distribute_iterator.cc:228]: we should improve type safety for iterators, in both definition and implementation. E.g the `GetKey` method of `WindowIterator` should mark as `const`

**Review by dl239 — APPROVED** [review]

File: src/catalog/distribute_iterator.cc [hunk 2]
```diff
@@ -179,70 +179,108 @@ void DistributeWindowIterator::Reset() {
...
+        if (it != nullptr) {
+            it->Seek(key);
+            if (it->Valid()) {
+                return {static_cast<int32_t>(pid), it, {}};
+            }
+            delete it;
+        }
+    }
+    DLOG(INFO) << "seek to key " << key << " from remote. "
+               << " cur_pid " << cur_pid_;
```
Comment (aceforeverd) [inline src/catalog/distribute_iterator.cc]: wrong log

File: src/catalog/distribute_iterator.cc [hunk 3]
```diff
@@ -179,70 +179,108 @@ void DistributeWindowIterator::Reset() {
...
+        std::string msg;
+        auto it = client_iter->second->Scan(tid_, pid, key, index_name_, 0, 0, FLAGS_traverse_cnt_limit, msg);
+        if (it != nullptr && it->Valid()) {
+            return {static_cast<int32_t>(pid), {}, it};
+        }
+    }
+
+    // fallback
+    for (const auto& kv : *tables_) {
+        if (kv.first <= cur_pid_) {
```
Comment (aceforeverd) [inline src/catalog/distribute_iterator.cc]: wrong compare

**Comment by aceforeverd** [comment]
[window union across tables](https://trello.com/c/9jpyUwAv/67-window-union-across-tables)

**Pull request merged**